Framework/Algorithms/src/SampleCorrections/MCInteractionVolume.cpp

```diff
@@ -90,7 +91,10 @@ int MCInteractionVolume::getComponentIndex(
   int componentIndex = -1;
   if (m_pointsIn != ScatteringPointVicinity::SAMPLEONLY && m_env) {
     const int randomStart =
-        (m_pointsIn == ScatteringPointVicinity::ENVIRONMENTONLY) ? 1 : 0;
+        ((m_pointsIn == ScatteringPointVicinity::ENVIRONMENTONLY) ||
```

Comment:
When editing this part, I had the feeling that it will be revisited one day :)
I think we should really streamline this. There is no point in generating random number if it has to be only 0 or only 1. Moreover, it is not explicit, that `nextInt` below generates in the interval inclusively. I am not sure actually what will happen it there is a multi-part container present. Could you double check?

Reply:
I investigated the behaviour of the nextInt call using the ISIS PEARL instrument which has 9 parts to its container. nextInt generates random integers including the start, end values supplied. So if m_PointsIn=SAMPLEANDENVIRONMENT the call to nextInt looks like:
```
    componentIndex =  rng.nextInt(0, 9) - 1;
```
The nextInt call returns numbers from 0 to 9 _including_ the two bounding integers. The componentIndex variable ends up being values between -1 and 8. The -1 value indicates the scatter point has been selected in the sample.

I tried out a few different ways of writing this and it seems there's a trade off between the code looking clean\concise and avoiding calling nextInt if the range contains only a single value. The implementation of the MersenneTwister random number generator has a shortcut in it if you supply it a range containing only one value where it returns the single value. See https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/inc/MantidKernel/uniform_int_distribution.h#L269. So maybe it's OK to call it with a single value range?

Here's one of my attempts at a rewrite that aims to make the code more simple to read - but it always calls nextInt. What do you think?
```
  const int sampleIndex = -1;
  const int firstEnvIndex = sampleIndex + 1;
  int randomStart = sampleIndex;
  int randomEnd = m_env ? static_cast<int>(m_env->nelements()) - 1 : sampleIndex;
  if ((m_pointsIn == ScatteringPointVicinity::ENVIRONMENTONLY) ||
      !m_sample->hasValidShape())
    randomStart = firstEnvIndex;
  if (m_pointsIn == ScatteringPointVicinity::SAMPLEONLY)
    randomEnd = sampleIndex;
  const int componentIndex = rng.nextInt(randomStart, randomEnd);
  return componentIndex;
```

Reply:
This looks better, I'd still suggest putting at the end:
```suggestion
if (randomStart == randomEnd) {
        return randomStart;
} else {
        // generates uniform random number in the closed interval [start,end]
        return rng.nextInt(randomStart, randomEnd);
 }
```

Reply:
I'd also not prefix the variables with random, just let them be start and end. It's only the last line that generates the random number in between.